### ISSUE TITLE
Data race condition in boost/asio/detail/scoped_lock.hpp

### ISSUE THREAD

**Issue by dimitrov-ds** [chriskohlhoff/asio#837]
Hi,
I've just run Synopsys Coverity, version 2020.03, static analysis on Boost v 1_75_0 
It has highlighted several potential issues one of which I will describe here.
In scoped_lock.hpp there is a mutex and a Boolean guard around the mutex. ‘locked_’ is checked prior to locking and unlocking the mutex and then set to true or false respectively.
This, to me and the Coverity static analysis, creates a data race condition. Is that a false positive in the static analysis tool? What am I missing?
The issue can be also reproduced by running Coverity static analysis on this boost example server:
https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/example/cpp03/ssl/server.cpp 

Thanks,
Dimitar

**Comment by gummif** [comment]
Yes, lock objects are not thread safe and the bool is to check whether unlocking is needed on destruction.
